### PR TITLE
feat: add ability to store current indexed data

### PR DESCRIPTION
**Pull request by azurwastaken**
# Pull Request type

- Feature

## What is the current behavior?

Resolves: #36 

## What is the new behavior?

The bot now have a StorageTrait and a StorageWrapper that allow him to handle storage of current state (indexed position and last witnessed block by the monitoring)

currently there is only one possibility which is dumping into a json file. the service will dump every tick updated last witnessed block and positions map into the json.

At the start the run_all_service will initiate a jsonStorage that will read the json file and get the data in it. 
if there is some data he will start the indexing service from the last synced block and the monitoring service's position map will be initialized with data from the json file.
Otherwise, we do as before.

## Does this introduce a breaking change?

No

## Other information

There is many TODO's related to this but as its for an mvp it look enough to me

### CODE REVIEW

File: src/storage/json.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+use serde::Serialize;
+use serde_json::Value;
+use std::collections::HashMap;
+
+use crate::types::position::{self, Position};
+
+use super::storage_wrapper::Storage;
+
+#[derive(Serialize, Default)]
+struct DataAsJson {
```
Comment (EvolveArt) [inline src/storage/json.rs:13]: how is that related to Json ?

File: src/storage/json.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        let map = DataAsJson {
+            last_block_indexed,
+            positions,
+        };
+        // Serialize the HashMap to a JSON string
+        let json = serde_json::to_string_pretty(&map)?;
+
+        // Write the JSON string to a file
+        let mut file = File::create(self.file_path.clone())?;
+        file.write_all(json.as_bytes())?;
```
Comment (EvolveArt) [inline src/storage/json.rs:96]: why do we write to json if we just write as bytes tho?

File: src/storage/storage_wrapper.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    fn load_state(&mut self) -> Result<(u64, HashMap<u64, Position>)>;
+    fn save_state(
+        &mut self,
+        position: HashMap<u64, position::Position>,
+        last_block_indexed: u64,
+    ) -> impl std::future::Future<Output = Result<()>> + Send;
+    fn get_last_saved_positions_map(&self) -> HashMap<u64, Position>;
+}
+
+pub enum StorageWrapper {
```
Comment (EvolveArt) [inline src/storage/storage_wrapper.rs:19]: remove the wrapper, we can just use the `Storage` trait

File: src/types/position.rs [hunk 4]
```diff
@@ -32,6 +33,10 @@ impl PositionsMap {
         Self(Arc::new(RwLock::new(HashMap::new())))
     }
 
+    pub fn from_backup(map_backup: HashMap<u64, Position>) -> Self {
```
Comment (EvolveArt) [inline src/types/position.rs:36]: `from_storage` might be a better name, or `from_local_storage`

**Review by EvolveArt — CHANGES_REQUESTED** [review]

File: src/storage/storage_wrapper.rs (hunk 3 above)
Comment (azurwastaken) [reply src/storage/storage_wrapper.rs]: the wrapper looked more clean and straight forward to me but done

File: src/types/position.rs (hunk 4 above)
Comment (azurwastaken) [reply src/types/position.rs]: done

File: src/storage/json.rs (hunk 2 above)
Comment (azurwastaken) [reply src/storage/json.rs]: i don't understand here, json is just a big String and we need to have it [u8] to write in file

File: src/storage/json.rs (hunk 1 above)
Comment (azurwastaken) [reply src/storage/json.rs:14]: it reflect json structure but maybe i can just rename it Data and put it in the storage.rs along the trait as it also represent the data we cant to store

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**